### PR TITLE
Add new automl seq2seq pytorch version

### PR DESCRIPTION
**Pull request by TheaperDeng**
- [x] Still need some UTs.
- [x] Need to be review for general hyperparams.

### CODE REVIEW

File: pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py [hunk 1]
```diff
@@ -16,187 +16,96 @@
...
+                                    lstm_layer_num,
+                                    dropout=dropout,
+                                    batch_first=True)
+        self.lstm_decoder = nn.LSTMCell(lstm_hidden_dim, lstm_hidden_dim)
+        fc_layers = [lstm_hidden_dim] + [fc_hidden_dim]*(fc_layer_num-1) + [output_feature_num]
+        layers = []
+        for i in range(fc_layer_num):
+            if i != 0:
+                layers += [nn.Dropout(p=dropout)]
+            layers += [nn.Linear(fc_layers[i], fc_layers[i+1])]
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py:46]: multiple linear layers without activation

File: pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py [hunk 2]
```diff
@@ -16,187 +16,96 @@
...
-        # decoder_input = [batch size, 1], since output_dim is 1
-        decoder_input = decoder_input.unsqueeze(1)
-        # decoder_input = [batch_size, 1, 1]
+        x, (hid, cta) = self.lstm_encoder(input_seq)
+        h, c = hid[-1], cta[-1]
 
-        output, (hidden, cell) = self.rnn(decoder_input, (hidden, cell))
+        decode_list = []
+        for i in range(self.output_step):
+            h, c = self.lstm_decoder(h, (h, c))
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py:56]: make sure decoder input is right

File: pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py [hunk 3]
```diff
@@ -16,187 +16,96 @@
...
-
-        # input = decoder_input.view(-1, 1)
-        # decoder_input = [batch size, 1], since output_dim is 1
-        decoder_input = decoder_input.unsqueeze(1)
-        # decoder_input = [batch_size, 1, 1]
+        x, (hid, cta) = self.lstm_encoder(input_seq)
+        h, c = hid[-1], cta[-1]
 
-        output, (hidden, cell) = self.rnn(decoder_input, (hidden, cell))
+        decode_list = []
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py:54]: initialize the output with `torch.zeros()` and update it step by step.

File: pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py [hunk 4]
```diff
@@ -16,187 +16,96 @@
...
+                                    dropout=dropout,
+                                    batch_first=True)
+        self.lstm_decoder = nn.LSTMCell(lstm_hidden_dim, lstm_hidden_dim)
+        fc_layers = [lstm_hidden_dim] + [fc_hidden_dim]*(fc_layer_num-1) + [output_feature_num]
+        layers = []
+        for i in range(fc_layer_num):
+            if i != 0:
+                layers += [nn.Dropout(p=dropout)]
+            layers += [nn.Linear(fc_layers[i], fc_layers[i+1])]
+        self.fc = nn.Sequential(*layers)
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py:47]: Is seq2seq a sequential?

File: pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py [hunk 5]
```diff
@@ -16,187 +16,96 @@
...
+class LSTMSeq2Seq(nn.Module):
+    def __init__(self,
+                 input_feature_num,
+                 future_seq_len,
+                 output_feature_num,
+                 lstm_hidden_dim=128,
+                 lstm_layer_num=2,
+                 fc_hidden_dim=128,
+                 fc_layer_num=2,
+                 dropout=0.25):
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py:33]: shall we use the same dropout for both encoder LSTM input and the fc layers in decoder or expose two dropout configurations for user?

File: pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py (hunk 2 above)
Comment (yushan111) [reply pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py:56]: maybe decoder should take different inputs in fit and predict?

File: pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py (hunk 1 above)
Comment (yushan111) [reply pyzoo/zoo/zouwu/model/forecast/model/Seq2Seq_pytorch.py:46]: If that is unnecessary, we could remove `fc_hidden_dim` and `fc_layer_num`

**Comment by TheaperDeng** [comment]
Three changes will be made
- [ ] Add option for teaching force training or not.
- [ ] Change the input of decoder as the last observed value rather than the encoder output.
- [ ] Use only 1 fc layer as final map (Reference: https://github.com/JEddy92/TimeSeries_Seq2Seq/blob/master/notebooks/TS_Seq2Seq_Intro.ipynb)

File: pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py [hunk 6]
```diff
@@ -16,187 +16,100 @@
...
-
-        # last hidden state of the encoder is used as the initial hidden state of the decoder
-        hidden, cell = self.encoder(source)
-
-        # Populate the first target sequence with end of encoding series value
-        # decoder_input : [batch_size, output_dim]
-        decoder_input = source[:, -1, :output_dim]
+    def forward(self, input_seq, target_seq=None):
+        x, (hidden, cell) = self.lstm_encoder(input_seq)
+        decoder_input = input_seq[:, -1, :] # last value
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py:52]: Do we need input_seq[:, -1, :output_dim]?

File: pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py [hunk 7]
```diff
@@ -16,187 +16,100 @@
...
+        decoder_output = torch.zeros(input_seq.shape[0], self.future_seq_len, self.output_feature_num)
         for i in range(self.future_seq_len):
-            decoder_output, hidden, cell = self.decoder(decoder_input, hidden, cell)
-            target_seq[:, i] = decoder_output
-            if target is None:
-                # in test mode
-                decoder_input = decoder_output
+            decoder_output_step, (hidden, cell) = self.lstm_decoder(decoder_input, (hidden, cell))
+            out_step = self.fc(decoder_output_step)
+            decoder_output[:,i:i+1,:] = out_step
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py:59]:
```
decoder_output[:,i] = out_step
```

**Review by yushan111 — COMMENTED** [review]
What is the difference between this version and the original [seq2seq pytorch](https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py) except the hyper-parameter names?

File: pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py [hunk 8]
```diff
@@ -16,187 +16,100 @@
...
-        # tensor to store the predicted outputs
-        target_seq = torch.zeros(batch_size, self.future_seq_len, output_dim)
-
-        # last hidden state of the encoder is used as the initial hidden state of the decoder
-        hidden, cell = self.encoder(source)
-
-        # Populate the first target sequence with end of encoding series value
-        # decoder_input : [batch_size, output_dim]
-        decoder_input = source[:, -1, :output_dim]
+    def forward(self, input_seq, target_seq=None):
```
Comment (yushan111) [inline pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py:50]: How to enable teacher forcing with input `target_seq`?

File: pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py (hunk 7 above)
Comment (TheaperDeng) [reply pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py:59]: > What is the difference between this version and the original [seq2seq pytorch](https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py) except the hyper-parameter names?

1. Add an option to choose teacher forcing or not
2. change `input_seq[:, -1, :output_dim]` to `input_seq[:, -1, :]` (actually I can not understand why we need `:output_dim`)
3. Add flex loss creator
4. Simplify the code

File: pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py (hunk 8 above)
Comment (TheaperDeng) [reply pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py:50]: teacher forcing is a config that can be set. it will expose to user just like other configs in Forecaster.

**Comment by TheaperDeng** [comment]
> What is the difference between this version and the original [seq2seq pytorch](https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py) except the hyper-parameter names?

1. Add an option to choose teacher forcing or not
2. change `input_seq[:, -1, :output_dim]` to `input_seq[:, -1, :]` (actually I can not understand why we need `:output_dim`)
3. Add flex loss creator
4. Simplify the code

File: pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py (hunk 8 above)
Comment (yushan111) [reply pyzoo/zoo/zouwu/model/Seq2Seq_pytorch.py:50]: yeah, I mean we need `target_seq` for teacher forcing, right? And how can we input `target_seq`

**Comment by yushan111** [comment]
> 2. change `input_seq[:, -1, :output_dim]` to `input_seq[:, -1, :]` (actually I can not understand why we need `:output_dim`, since it is possible that output_dim>input_dim. Such that we need to predict 5 different KPIs with only 4 features.)

Maybe that is to handle input_dim > output_dim. Such as input features of [`target_value_1`, `target_value_2`, `extra_feature_1`, `extra_feature_2`] and output target of [`target_value_1`, `target_value_2`]. Maybe it doesn't work if the input feature order changes (target values are not in the first), but we used to assume model takes the output of feature transformer. We could make it more flexible.

Also we need tests of  `input_feature_num` > `output_feature_num`

**Comment by TheaperDeng** [comment]
PR validation combine pass:
http://10.239.47.210:18888/job/ZOO-PR-Python-ExampleTests-py36-ray/1592/
http://10.239.47.210:18888/job/ZOO-PR-Validation/5343/
